tests/Config/Plugin/Hook/CustomArrayMapStorageProvider.php

```diff
@@ -0,0 +1,177 @@
+<?php
+
+namespace Psalm\Tests\Config\Plugin\Hook;
+
+use Psalm\Codebase;
+use Psalm\Plugin\DynamicFunctionStorage;
+use Psalm\Plugin\DynamicTemplateProvider;
+use Psalm\Plugin\EventHandler\Event\FunctionDynamicStorageProviderEvent;
+use Psalm\Plugin\EventHandler\FunctionDynamicStorageProviderInterface;
+use Psalm\Storage\FunctionLikeParameter;
+use Psalm\Type;
+use Psalm\Type\Atomic\TCallable;
+use Psalm\Type\Atomic\TTemplateParam;
+use Psalm\Type\Union;
+
+use function array_keys;
+use function array_map;
+use function count;
+
+class CustomArrayMapStorageProvider implements FunctionDynamicStorageProviderInterface
+{
+    public static function getFunctionIds(): array
+    {
+        return ['custom_array_map'];
+    }
+
+    public static function getFunctionStorage(FunctionDynamicStorageProviderEvent $event): ?DynamicFunctionStorage
+    {
+        $func_call = $event->getExpr();
+
+        if ($func_call->isFirstClassCallable()) {
```

Comment:
This must probably be included in every plugin?
Would it make sense to run any dynamic function providers for first class callables?

Reply:
> Would it make sense to run any dynamic function providers for first class callables?

What do you think of this case?

```php
/** @var list<int> */
$numbers = [1, 2, 3];

pipe(
    $numbers,
    partial_left(custom_array_map(...), fn($i) => $i + 1)
)
```
Theoretically, it is possible to handle this situation.

Reply:
There are some issues here:

`custom_array_map(....)` results in a FunctionCall with first argument of `VariadicPlaceholder`.
The result of this is a closure - of which the type can be determined by the expression analyzer.
This means that the `FunctionDynamicStorageProviderEvent` is triggered for the code `custom_array_map(...)` - but you cannot really do anything on that.


The second issue here, is that there is no support for templates on closures inside psalm yet.
See https://github.com/vimeo/psalm/issues/7244

Since first class callable is just a wrapper around `Closure::fromCallable()` - it means that templates get lost.
See:
https://psalm.dev/r/01e2acbe39


If psalm were able to determine that, in this case, debug returns `Closure(T): T` instead of `Closure(mixed): mixed`, there would never be a need for adding a `FunctionDynamicStorageProviderEvent`.

Since first class callable is going to be used quite often in the future, this feels like a missing part in psalm.
Also scrolled trhough the issues and saw that it is not supported for reasons and that psalm even reports an issue if you add templates to closure docblocks:

"Templated closures are not supported"

https://psalm.dev/r/2da9ff0b52

I think with first class callables, it is a feature that should be working.
More info:

- https://github.com/vimeo/psalm/issues/5472
- https://github.com/vimeo/psalm/issues/5471#issuecomment-805830527

Reply:
Ow snap .... on second thought - in this case `custom_array_map()` is dealt with by a plugin as well.
This indeed indicates that we need it to be able to determine a return type in `FunctionDynamicStorageProviderEvent`.

What would you let it return in this case?

Reply:
I've removed first class callable support from hook for now.
Current issue is getting bigger and bigger. Let's deal with this in another issue.

Reply:
Alright, moved over this info to #7244, we can pick it up in another pr / issue.

If you don't trigger the plugin for first class callables, it will fall back to the signature in the dockblocks on the function.
Which will propbably look like this:

```php
/**
 * @param list<mixed> $args
 * @return list<mixed>
 */
function custom_array_map(...$args) {
    //...
}
```

It is ok, but not very usefull for partial function application though.
I can't come up with how you would improve the type from within the plugin - of that's even ever possible.
Because it requires the context of the parent function to see how the of will be used. 
(Basically you want to run the plugin on the result of the first class callable instead of on the function call)